book/src/proposals/validator-timestamp-oracle.md

```diff
@@ -0,0 +1,118 @@
+law enforcement. This proposal describes a validator timestamp oracle that
+would allow a Solana cluster to satisfy this need.
+
+The general outline of the proposed implementation is as follows:
+
+- At regular intervals, each validator records its observed time for a known slot
+  on-chain (via a Timestamp added to a slot Vote)
+- A client can request a block time using the `getBlockTime` RPC method. When a
+  client requests a timestamp for block N:
+
+  1. A validator determines a "cluster" timestamp for a recent slot before block N
+  by observing all the Timestamp instructions recorded on the ledger that
+  reference that slot, and determining the stake-weighted mean timestamp.
+
+  2. This recent mean timestamp is then used to calculate the timestamp of
+  block N using the cluster's established slot duration
+
+Requirements:
+- Any validator replaying the ledger in the future must come up with the same
+  time for every block since genesis
+- Estimated block times should not drift more than an hour or so before resolving
+  to real-world (oracle) data
+- The block times are not controlled by a single centralized oracle, but
+  ideally based on a function that uses inputs from all validators
+- Each validator must maintain a timestamp oracle
+
+## Recording Time
+
+At regular intervals, each validator records its observed time by submitting a
+VoteWithTimestamp instruction to the cluster as it is voting on a particular
+slot. In addition to the normal Vote data, the instruction contains the slot
+index and the validator's system time as Unix timestamp (seconds since the Unix
+epoch). It is signed by the validator's identity keypair as a usual Vote.
+
+This proposal suggests that VoteWithTimestamp instructions be issued every
+30min, which should be short enough to prevent block times drifting very much,
+without adding too much transaction overhead to the cluster. Validators can
+convert this time to a slot interval using the `slots_per_year` value that is
+stored in each bank.
+
+```text
+let seconds_in_30min = 1800;
+let timestamp_interval = (slots_per_year / SECONDS_PER_YEAR) * seconds_in_30min;
+```
+
+The VoteWithTimestamp variation should be triggered in `replay_stage::handle_votable_bank()`
+when `bank.slot() % timestamp_interval == 0`.
+
+### Vote Accounts
+
+A validator's vote account will hold its most recent slot-timestamp in VoteState.
+
+### Vote Program
+
+The on-chain Vote program needs to be extended to process a VoteWithTimestamp
+instruction from validators. In addition to its current process_vote
+functionality (including loading the correct Vote account and verifying that the
+transaction signer is the expected validator), this process needs to compare the
+slot and timestamp to the currently stored values to verify that they are both
+monotonically increasing, and store the new slot and timestamp in the account.
+
+## Calculating Median Timestamp
```

Comment:
Instead of "Median", how about "Stake-weighted Mean"

Reply:
🤦‍♀